crates/stages/src/sets.rs

```diff
@@ -0,0 +1,167 @@
+//! // Build a pipeline with all offline stages and a custom stage at the end.
+//! Pipeline::builder()
+//!     .add_stages(
+//!         OfflineStages.build().add_stage(MyCustomStage)
+//!     )
+//!     .build();
+//! ```
+use crate::{
+    stages::{
+        AccountHashingStage, BodyStage, ExecutionStage, HeaderStage, MerkleStage,
+        SenderRecoveryStage, StorageHashingStage, TotalDifficultyStage,
+    },
+    StageSet, StageSetBuilder,
+};
+use reth_db::database::Database;
+use reth_interfaces::{
+    consensus::Consensus,
+    p2p::{bodies::downloader::BodyDownloader, headers::downloader::HeaderDownloader},
+};
+use reth_primitives::ChainSpec;
+use std::sync::Arc;
+
+/// A set containing all stages to run a fully syncing instance of reth.
+///
+/// A combination of (in order)
+///
+/// - [`OnlineStages`]
+/// - [`OfflineStages`]
+#[derive(Debug)]
+pub struct DefaultStages<H, B> {
+    /// Configuration for the online stages
+    online: OnlineStages<H, B>,
+}
+
+impl<H, B> DefaultStages<H, B> {
+    /// Create a new set of default stages with default values.
+    pub fn new(consensus: Arc<dyn Consensus>, header_downloader: H, body_downloader: B) -> Self {
+        Self { online: OnlineStages::new(consensus, header_downloader, body_downloader) }
+    }
+}
+
+impl<DB, H, B> StageSet<DB> for DefaultStages<H, B>
+where
+    DB: Database,
+    H: HeaderDownloader + 'static,
+    B: BodyDownloader + 'static,
+{
+    fn build(self) -> StageSetBuilder<DB> {
+        self.online.build().add_set(OfflineStages)
+    }
+}
+
+/// A set containing all stages that require network access by default.
+///
+/// These stages *can* be run without network access if the specified downloaders are
+/// themselves offline.
+#[derive(Debug)]
+pub struct OnlineStages<H, B> {
+    /// The consensus engine used to validate incoming data.
+    consensus: Arc<dyn Consensus>,
+    /// The block header downloader
+    header_downloader: H,
+    /// The block body downloader
+    body_downloader: B,
+}
+
+impl<H, B> OnlineStages<H, B> {
+    /// Create a new set of online stages with default values.
+    pub fn new(consensus: Arc<dyn Consensus>, header_downloader: H, body_downloader: B) -> Self {
+        Self { consensus, header_downloader, body_downloader }
+    }
+}
+
+impl<DB, H, B> StageSet<DB> for OnlineStages<H, B>
+where
+    DB: Database,
+    H: HeaderDownloader + 'static,
+    B: BodyDownloader + 'static,
+{
+    fn build(self) -> StageSetBuilder<DB> {
+        StageSetBuilder::default()
+            .add_stage(HeaderStage::new(self.header_downloader, self.consensus.clone()))
+            .add_stage(TotalDifficultyStage::default())
+            .add_stage(BodyStage { downloader: self.body_downloader, consensus: self.consensus })
+    }
+}
+
+/// A set containing all stages that do not require network access.
+///
+/// A combination of (in order)
+///
+/// - [`ExecutionStages`]
+/// - [`HashingStages`]
+#[derive(Debug)]
+pub struct OfflineStages;
```

Comment:
is this still unimplemented or just used as an entrypoint to build a new StageSet?

perhaps derive Default, mark non_exhaustive?

Reply:
It is a stageset in itself, it does not hold any stage since you can override config using `OfflineStages.build().set(stage)`. Can mark `non_exhaustive`

Reply:
+1, mark non_exhaustive, and let's add a Default impl to instantiate